waywiser_node/src/waywise_truck.cpp

```diff
@@ -0,0 +1,554 @@
+                                                                                   // description is VESC
+        mVESCMotorController->connectSerial(portInfo);
+        RCLCPP_INFO(
+          get_logger(), "VESCMotorController connected to: %s",
+          portInfo.systemLocation().toLocal8Bit().data());
+      }
+    }
+
+    if (mVESCMotorController->isSerialConnected()) {
+      mCarMovementController->setMotorController(mVESCMotorController);
+      mVESCMotorController->setPollValuesPeriod(mUpdateVehicleStatePeriod.count());
+
+      // VESC is a special case that can also control the servo
+      const auto servoController = mVESCMotorController->getServoController();
+      servoController->setInvertOutput(invert_servo_output_);
+      servoController->setServoRange(servo_max_ - servo_min_);
+      servoController->setServoCenter(servo_offset_);
+      mCarMovementController->setServoController(servoController);
+
+      is_in_simulation_mode_ = false;
+      waywise_posType_used_ = PosType::fused;
+    } else {
+      // publish periodically with timer when no motorcontroller connected
+      // (simulation)
+      is_in_simulation_mode_ = true;
+      waywise_posType_used_ = PosType::simulated;
+      simulation_timer_ =
+        this->create_wall_timer(
+        mUpdateVehicleStatePeriod,
+        std::bind(&WayWiseTruck::simulation_timer_callback, this));
+
+      RCLCPP_INFO(
+        get_logger(),
+        "VESCMotorController is not connected. "
+        "waywise_truck is in simulation mode!");
+    }
+
+    // --- Positioning setup ---
+    // Position Fuser
+    positionFuser = new SDVPVehiclePositionFuser(this);
+    // GNSS (with fused IMU when using u-blox F9R)
+    mUbloxRover.reset(new UbloxRover(mTruckState));
+    foreach(const QSerialPortInfo & portInfo, QSerialPortInfo::availablePorts()) {
+      // qDebug()<<portInfo.manufacturer();
+      if (portInfo.manufacturer().toLower().replace("-", "").contains("ublox")) {
+        if (mUbloxRover->connectSerial(portInfo)) {
+          qDebug() << "UbloxRover connected to:" << portInfo.systemLocation();
+
+          //mUbloxRover->setIMUOrientationOffset(0.0, 0.0, 0.0);
+        }
+      }
+    }
+
+    rtcmClient = new RtcmClient(this);
+    QObject::connect(
+      mUbloxRover.get(), &UbloxRover::updatedGNSSPositionAndYaw, positionFuser,
+      &SDVPVehiclePositionFuser::correctPositionAndYawGNSS);
+
+    // -- NTRIP/TCP client setup for feeding RTCM data into GNSS receiver
+    QObject::connect(
+      mUbloxRover.get(), &UbloxRover::gotNmeaGga, rtcmClient, &RtcmClient::forwardNmeaGgaToServer);
+    QObject::connect(
+      rtcmClient, &RtcmClient::rtcmData,
+      mUbloxRover.get(), &UbloxRover::writeRtcmToUblox);
+    QObject::connect(
+      rtcmClient, &RtcmClient::baseStationPosition,
+      mUbloxRover.get(), &UbloxRover::setEnuRef);
+    if (rtcmClient->connectWithInfoFromFile("./rtcmServerInfo.txt")) {
```

Comment:
I am not sure what the "ROS2 way" is to save credentials (that you do not want to end up in git) but want to mention that connectWithInfoFromFile(..) calls connectNtrip(serverName, port, ntripConnectionInfo) internally so it does not *need* to be a file.
See: https://github.com/RISE-Dependable-Transport-Systems/WayWise/blob/4ffa73b76e2d882efe3f8bc6eeeb4162b05f3883/sensors/gnss/rtcmclient.cpp#L82

Reply:
Good point, I think it's still better to read credentials from a file stored outside the repository. With "./rtcmServerInfo.txt" filepath here, by launching the node from the root directory of workspace, the node looks for the file in that dir.